File: cut-off-trees-for-golf-event.py
# https://leetcode.com/problems/cut-off-trees-for-golf-event/description/
# 1. Sort by numbers
# 2. BFS 
# Time Limit Exceeded (needs to change 2. BFS -> Hadlock's Algorithm)
class Solution:
    def cutOffTree(self, forest):
        l1 = len(forest)
        if l1==0: return 0
        l2 = len(forest[0])
        if l2==0: return 0
        if forest[0][0]==0: return -1
        p = {}
        for i in range(l1):
            for j in range(l2):
                if forest[i][j]!=0: p[forest[i][j]]=[i,j]
        pKeys = sorted(p.keys())
        print(len(pKeys))
        ans = 0
        preP = [0,0]
        for k in range(len(pKeys)):
            curP=p[pKeys[k]]
            print(curP)
            if forest[curP[0]][curP[1]]==0: return -1
            visited = [[False for j in range(l2)] for i in range(l1)]
            q = collections.deque()
            q.append([preP[0],preP[1],0])
            curCnt = -1
            while q:
                cur = q.popleft()
                i,j,cnt = cur[0],cur[1],cur[2]
                #print(i,j,cnt,visited[i][j])
                if not visited[i][j]:
                    visited[i][j]=True
                    if i==curP[0] and j==curP[1]:
                        curCnt=cnt
                        break
                    else:
                        if i>0 and not visited[i-1][j] and forest[i-1][j]!=0: q.append([i-1,j,cnt+1])
                        if j>0 and not visited[i][j-1] and forest[i][j-1]!=0: q.append([i,j-1,cnt+1])
                        if i<l1-1 and not visited[i+1][j] and forest[i+1][j]!=0: q.append([i+1,j,cnt+1])
                        if j<l2-1 and not visited[i][j+1] and forest[i][j+1]!=0: q.append([i,j+1,cnt+1])
            if curCnt==-1: return -1
            else:
                ans+=curCnt
                preP=curP
        return ans
        

File: cut-off-trees-for-golf-event.java
// https://leetcode.com/problems/cut-off-trees-for-golf-event/description/
// 1. Sort by numbers
// 2. BFS
// Time: O(n^2), Space: O(n)
class Solution {
    public int cutOffTree(List<List<Integer>> forest) {
        int l1 = forest.size();
        if(l1==0) return 0;
        int l2 = forest.get(0).size();
        if(l2==0) return 0;
        
        TreeMap<Integer,Integer> pMap = new TreeMap<Integer,Integer>();
        for(int i=0; i<l1; i++){
            for(int j=0; j<l2; j++){
                if(forest.get(i).get(j)>1){
                    pMap.put(forest.get(i).get(j), i*l2+j);
                }
            }
        }
        
        int ans=0, pre=0;
        if(forest.get(0).get(0)==0) return -1;
        boolean visit[][] = new boolean[l1][l2];
        LinkedList<Integer> q = new LinkedList<Integer>();
        for(Map.Entry<Integer,Integer> e : pMap.entrySet()){
            int cur = e.getValue();
            int ei=cur/l2, ej=cur%l2;
            if(forest.get(ei).get(ej)==0) return -1;
            q.clear();
            q.add(pre);
            for(boolean[] v : visit) Arrays.fill(v,false);
            int cnt=0,curCnt=-1;
            while(!q.isEmpty()){
                int size=q.size();
                //System.out.println(size);
                for(int k=0; k<size; k++){
                    int p = q.poll();
                    int i=p/l2, j=p%l2;
                    //System.out.println(i+","+j);
                    if(i==ei && j==ej){
                        curCnt=cnt;
                        break;
                    }else{
                        visit[i][j]=true;
                        if(i>0 && forest.get(i-1).get(j)!=0 && !visit[i-1][j]){ q.add((i-1)*l2+j); visit[i-1][j]=true; }
                        if(j>0 && forest.get(i).get(j-1)!=0 && !visit[i][j-1]){ q.add(i*l2+j-1); visit[i][j-1]=true; }
                        if(i<l1-1 && forest.get(i+1).get(j)!=0 && !visit[i+1][j]){ q.add((i+1)*l2+j); visit[i+1][j]=true; }
                        if(j<l2-1 && forest.get(i).get(j+1)!=0 && !visit[i][j+1]){ q.add(i*l2+j+1); visit[i][j+1]=true; }
                    }
                }
                if(curCnt>-1) break;
                cnt++;
            }
            if(curCnt==-1) return -1;
            ans+=curCnt;
            pre=cur;
        }
        
        return ans;
    }
}
